util/which-key/which-key.lisp

```diff
@@ -0,0 +1,45 @@
+;;;; which-key.lisp
+
+(in-package #:which-key)
+
+(export '(get-kmaps-at-key-seq key-press-hook replace-hook))
+
+;;; "which-key" goes here. Hacks and glory await!
+
+(defun get-kmaps-at-key (kmaps key)
+  (dereference-kmaps
+   (reduce
+    (lambda (result map)
+      (let* ((binding (find key (kmap-bindings map)
+                            :key 'binding-key :test 'equalp))
+             (command (when binding (binding-command binding))))
+        (if command
+            (setf result (cons command result))
+            result)))
+    kmaps
+    :initial-value '())))
+
+(defun get-kmaps-at-key-seq (kmaps key-seq)
+  "get a list of kmaps that are activated when pressing KEY-SEQ when
+KMAPS are enabled"
+  (if (= 1 (length key-seq))
+      (get-kmaps-at-key kmaps (first key-seq))
+      (get-kmaps-at-key-seq (get-kmaps-at-key kmaps (first key-seq))
+                            (rest key-seq))))
+
+(defun key-press-hook (key key-seq cmd)
+  "which-key replacement for *key-press-hook*"
+  (declare (ignore key))
+  (unless (eq *top-map* *resize-map*)
+    (let ((maps (get-kmaps-at-key-seq (dereference-kmaps (top-maps))
+                                        (reverse key-seq))))
+      (when (remove-if-not 'kmap-p maps)
+        (apply 'display-bindings-for-keymaps (reverse (cdr key-seq)) maps)))))
+
+(defmacro replace-hook (hook fn)
```

Comment:
This macro only expands to `add-hook`, the `remove-hook` form is discarded as it is not returned by the macro. If you want to sequence them you have to wrap them in a `progn` form. Furthermore you are evaluating the arguments of the macro multiple times. You can work around this by using `alexandria:once-only` or writting the gensyms yourself. I prefer using alexandria but you may not want to add a dependency, if so go with the vanilla version

### Vanilla CL

```lisp
(defmacro replace-hook (hook fn)
  "Easily replace a hook with a new function"
  (let ((!fn (gensym))
        (!hook (gensym)))
    `(let ((,!fn ,fn)
           (,!hook ,hook))
       (progn (remove-hook ,!hook ,!fn)
              (add-hook ,!hook ,!fn)))))
```

### Using Alexandria
```lisp
(defmacro replace-hook (hook fn)
  "Easily replace a hook with a new function"
  (alexandria:once-only (fn hook)
    `(progn (remove-hook ,!hook ,!fn)
            (add-hook ,!hook ,!fn))))
```